File: runtime/tools/attention.py
"""Attention and look-at tools (soft takeover)."""
from __future__ import annotations

from typing import Any, Dict

from runtime.schemas import ToolResult
from runtime.tools.registry import ToolContext, ToolRegistry, ToolSpec


ENTITY_SCHEMA: Dict[str, Any] = {
    "type": "object",
    "properties": {"entity": {"type": "string"}},
    "required": ["entity"],
    "additionalProperties": False,
}

EMPTY_SCHEMA: Dict[str, Any] = {"type": "object", "properties": {}, "additionalProperties": False}


def look_at(context: ToolContext, entity: str) -> ToolResult:
    return context.dispatch_action("look_at", {"entity": entity})


def face(context: ToolContext, entity: str) -> ToolResult:
    return context.dispatch_action("face", {"entity": entity})


def clear_attention(context: ToolContext) -> ToolResult:
    return context.dispatch_action("clear_attention", {})


def register_attention_tools(registry: ToolRegistry) -> None:
    registry.register(ToolSpec(
        "look_at",
        "Look at an entity without preventing other actions.",
        ENTITY_SCHEMA,
        look_at,
        "attention",
    ))
    registry.register(ToolSpec(
        "face",
        "Turn the body/head toward an entity.",
        ENTITY_SCHEMA,
        face,
        "attention",
    ))
    registry.register(ToolSpec(
        "clear_attention",
        "Stop looking/attending to the current target.",
        EMPTY_SCHEMA,
        clear_attention,
        "attention",
    ))

File: runtime-go/internal/tools/attention.go
package tools

import "github.com/ThanabordeeN/AgentRD2/runtime-go/internal/domain"

// attentionEntitySchema mirrors “ENTITY_SCHEMA“ in “attention.py“.
var attentionEntitySchema = map[string]any{
	"type": "object",
	"properties": map[string]any{
		"entity": map[string]any{"type": "string"},
	},
	"required":             []any{"entity"},
	"additionalProperties": false,
}

// attentionEmptySchema mirrors “EMPTY_SCHEMA“ in “attention.py“.
var attentionEmptySchema = map[string]any{
	"type":                 "object",
	"properties":           map[string]any{},
	"additionalProperties": false,
}

// lookAt mirrors “runtime/tools/attention.py::look_at“. The agent layer
// defaults a missing entity to "player".
func lookAt(ctx *Context, args map[string]any) domain.ToolResult {
	entity, ok := stringArg(args, "entity")
	if !ok {
		entity = "player"
	}
	return ctx.DispatchAction("look_at", map[string]any{"entity": entity})
}

// face mirrors “face“.
func face(ctx *Context, args map[string]any) domain.ToolResult {
	entity, ok := stringArg(args, "entity")
	if !ok {
		entity = "player"
	}
	return ctx.DispatchAction("face", map[string]any{"entity": entity})
}

// clearAttention mirrors “clear_attention“.
func clearAttention(ctx *Context, args map[string]any) domain.ToolResult {
	return ctx.DispatchAction("clear_attention", map[string]any{})
}

// RegisterAttentionTools registers the attention family (3 tools).
func RegisterAttentionTools(registry *Registry) {
	_ = registry.Register(Spec{
		Name:        "look_at",
		Description: "Look at an entity without preventing other actions.",
		Parameters:  attentionEntitySchema,
		Category:    "attention",
		Handler:     lookAt,
	})
	_ = registry.Register(Spec{
		Name:        "face",
		Description: "Turn the body/head toward an entity.",
		Parameters:  attentionEntitySchema,
		Category:    "attention",
		Handler:     face,
	})
	_ = registry.Register(Spec{
		Name:        "clear_attention",
		Description: "Stop looking/attending to the current target.",
		Parameters:  attentionEmptySchema,
		Category:    "attention",
		Handler:     clearAttention,
	})
}
